### PR TITLE
TRT-1854: translate test annotations to use: `include`/`exclude` CEL, labels, filtering

### PR DESCRIPTION
**Pull request by smg247**
`openshift-tests-extension` introduces a new way to skip upstream tests based on environment flags and `include`/`exclude` fields present on the listed tests. This PR translates the existing annotations to that functionality, but leaves them in place. They will be removed in a later PR.

### CODE REVIEW

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go [hunk 1]
```diff
@@ -97,12 +97,153 @@ func main() {
...
+		et.NameContainsAll("[sig-network] LoadBalancers [Feature:LoadBalancer]", "session affinity"),
+	}).Exclude(et.PlatformEquals("aws"))
+
+	specs.SelectAny([]et.SelectFunction{
+		et.NameContains("Networking should provide Internet connection for containers"), // Azure does not allow ICMP traffic to internet.
+		// Azure CSI migration changed how we treat regions without zones.
+		// See https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=2066865
+		et.NameContains("[sig-storage] In-tree Volumes [Driver: azure-disk] [Testpattern: Dynamic PV (immediate binding)] topology should provision a volume and schedule a pod with AllowedTopologies"),
+		et.NameContains("[sig-storage] In-tree Volumes [Driver: azure-disk] [Testpattern: Dynamic PV (delayed binding)] topology should provision a volume and schedule a pod with AllowedTopologies"),
+	}).Exclude(et.PlatformEquals("azure"))
```
Comment (stbenjam) [inline openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:139]: Do you think it would make sense to have for the simple cases an abstraction layer on top of this?

```
specs = filterByPlatform(specs)
```

And then in there have something like this:

```go
var platformExclusions = map[string][]string{
    "azure": {
        "Networking should provide Internet connection for containers", // Azure does not allow ICMP traffic to internet.
        "[sig-storage] In-tree Volumes [Driver: azure-disk] [Testpattern: Dynamic PV (immediate binding)] topology should provision a volume and schedule a pod with AllowedTopologies",
        "[sig-storage] In-tree Volumes [Driver: azure-disk] [Testpattern: Dynamic PV (delayed binding)] topology should provision a volume and schedule a pod with AllowedTopologies",
    },
   [....]
}

// iterate over the exclusions here
```

I'm just trying to imagine a more maintainable (and easier to modify) list for the simple cases.

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go [hunk 2]
```diff
@@ -97,12 +97,54 @@ func main() {
...
+	filterByTopology(specs)
+
+	// Tests which can't be run/don't make sense to run against a cluster with all optional capabilities disabled
+	specs.SelectAny([]et.SelectFunction{
+		// Requires CSISnapshot capability
+		et.NameContains("[Feature:VolumeSnapshotDataSource]"),
+		// Requires Storage capability
+		et.NameContains("[Driver: aws]"),
+		et.NameContains("[Feature:StorageProvider]"),
+	}).Exclude(et.NoOptionalCapabilitiesExist())
```
Comment (bertinatto) [inline openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:140]: This new capability in OTE is definitely a nice addition, but I also find it considerably more confusing compared to our current approach.

For example, compare the block above with our existing method:

```
// Tests which can't be run/don't make sense to run against a cluster with all optional capabilities disabled
"[Skipped:NoOptionalCapabilities]": {
    // Requires CSISnapshot capability
    `\[Feature:VolumeSnapshotDataSource\]`,
    // Requires Storage capability
    `\[Driver: aws\]`,
    `\[Feature:StorageProvider\]`,
},
```

Right now, adding a quick test skip is straightforward, but achieving the same result with OTE feels significantly more complex. I can imagine situations (like during a kube rebase) where we need to move quickly and tweak skips on the fly. Having to create a rule in OTE at that moment could slow things down and add friction (specially if the engineer is new to this framework). Not to mention that the rules we have in this PR are just a small subset of all the rules that would need to be migrated.

Perhaps the reason this feels more confusing is that `rules.go` already contains so many "twisted" skips and corner cases.

Given that, I’m tempted to suggest that we hold off on using this OTE capability in `k8s-tests.go` for now. It might be beneficial to first see it in action with other external binaries before making this transition.

Thoughts?

Comment (stbenjam) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:140]: I think moving this to a separate file, maybe called rules.go, would help keep the main code of k8s-tests-ext cleaner by separating concerns.

Regarding the semantics, while this approach is _different_, I don't think it's more difficult. In fact, I would argue it's simpler than the current use of annotations. With this method, you don't need to invoke any make targets to regenerate annotations—it takes effect immediately. You can also test it easily by invoking k8s-tests-ext with the appropriate environment flags.  Want to know what tests are running on AWS? `k8s-tests-ext list tests --platform=aws`.

Most importantly, test names remain unchanged. This is crucial because Kubernetes tests are the most volatile, and we have to maintain a history of tests across releases. The upcoming introduction of two-node OpenShift is likely to cause some test name churn. While ci-test-mapping can handle renames, our track record for keeping that updated isn't great.

Additionally, the language used in this approach is more consistent compared to annotations, and it provides access to a broader range of cluster information (and it's easy to expose new information). 

Initially, I had some concerns about how this would look in practice, but that led to helpers like filterByPlatform below. For most simple skip conditions, you would simply insert a record into the slice, making it straightforward to manage.

Do you have a recent example that illustrates why this path might be more challenging? That could help us address any potential concerns more directly.

Comment (bertinatto) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:140]: > Most importantly, test names remain unchanged. This is crucial because Kubernetes tests are the most volatile, and we have to maintain a history of tests across releases. The upcoming introduction of two-node OpenShift is likely to cause some test name churn. While ci-test-mapping can handle renames, our track record for keeping that updated isn't great.

Could you explain how this new approach will prevent us from maintaining a history of test names across releases? For instance, let's say a set of upstream tests have their prefix renamed from `[Feature:SELinuxMount]` to `[FeatureGate:SELinuxMount]`. We'd need to update the rule here as well, correct? I'm not doubting your point, just asking for my understanding.

> Do you have a recent example that illustrates why this path might be more challenging? That could help us address any potential concerns more directly.

IMO the `[Feature:LoadBalancer]` skips above are a good example. I keep staring at it, but still don't know how I'd create these rules myself. :sweat_smile: But maybe that's just me.

Comment (bertinatto) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:140]: Another immediate problem that I see is our `k8s-*` jobs in o/k. Those jobs use `ginkgo` to run the o/k tests and they rely on test annotations (some of which we add in `rules.go`), to skip some incompatible tests.

https://github.com/openshift/kubernetes/blob/master/openshift-hack/test-kubernetes-e2e.sh#L28

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go [hunk 3]
```diff
@@ -142,3 +184,137 @@ func convertToImage(obj interface{}) extension.Image {
...
+
+	for topology, exclusions := range topologyExclusions {
+		var selectFunctions []et.SelectFunction
+		for _, exclusion := range exclusions {
+			selectFunctions = append(selectFunctions, et.NameContains(exclusion))
+		}
+
+		specs.SelectAny(selectFunctions).Exclude(et.TopologyEquals(topology))
+	}
+}
```
[the remark this answers is not in the document]
Comment (smg247) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:320]: When running the `list tests` command on the binary, the tests will include an `environmentSelector` field that will display the conditions they are skipped under

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go [hunk 4]
```diff
@@ -97,12 +97,54 @@ func main() {
...
+		Exclude(et.Or(
+			et.PlatformEquals("alibabacloud"),
+			et.PlatformEquals("baremetal"),
+			et.PlatformEquals("ibmcloud"),
+			et.PlatformEquals("kubevirt"),
+			et.PlatformEquals("nutanix"),
+			et.PlatformEquals("openstack"),
+			et.PlatformEquals("ovirt"),
+			et.PlatformEquals("vsphere"),
+		))
```
[the remark this answers is not in the document]
Comment (smg247) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:120]: As it is, it will linger on until it is removed by someone. 
We could consider having some sort of a `MustSelect` function that errors if it doesn't find anything, but that doesn't currently exist.

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go (hunk 2 above)
Comment (stbenjam) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:140]: >> Most importantly, test names remain unchanged. This is crucial because Kubernetes tests are the most volatile, and we have to maintain a history of tests across releases. The upcoming introduction of two-node OpenShift is likely to cause some test name churn. While ci-test-mapping can handle renames, our track record for keeping that updated isn't great.

> Could you explain how this new approach will prevent us from maintaining a history of test names across releases? For instance, let's say a set of upstream tests have their prefix renamed from [Feature:SELinuxMount] to [FeatureGate:SELinuxMount]. We'd need to update the rule here as well, correct? I'm not doubting your point, just asking for my understanding.

It doesn't eliminate all renames.  It does eliminate altering a test name to change where a test runs. You would not update something here in the example you gave. There's no current plans to handle feature gates this way, but you could control execution of that FeatureGate if for say it didn't work on gcp and you'd be able to do that without renaming the test.

A reminder the process for single node disabling an upstream test, Egli had to:

1. Open a PR to o/k https://github.com/openshift/kubernetes/pull/2033 - which involved updating rules.go, and running the make target
2. Revendor that code in origin https://github.com/openshift/origin/pull/28948
3. Update ci-test-mapping https://github.com/openshift-eng/ci-test-mapping/pull/151

The process here is:

1. Add the test to the single-node topology exclusions list


>> Do you have a recent example that illustrates why this path might be more challenging? That could help us address any potential concerns more directly.

>IMO the [Feature:LoadBalancer] skips above are a good example. I keep staring at it, but still don't know how I'd create these rules myself. 😅 But maybe that's just me.

The functions being called are helpers provided by OTE, you don't have to write any. If you had something similar you would just duplicate this block.  If you wanted to also exclude the LoadBalancer tests on AWS, you'd just add et.PlatformEquals("aws") to the list.

If there's examples that are missing we can add more, but I don't think this is significantly different effort than maintaining the annotations.

>> Another immediate problem that I see is our k8s-* jobs in o/k. Those jobs use ginkgo to run the o/k tests and they rely on test annotations (some of which we add in rules.go), to skip some incompatible tests.
>> https://github.com/openshift/kubernetes/blob/master/openshift-hack/test-kubernetes-e2e.sh#L28

What are these jobs used for? Is it even needed anymore with extensions? If you open a kube bump PR, any new tests are automatically run in the presubmits so why do we need this separate script?

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go [hunk 5]
```diff
@@ -13,7 +13,7 @@ import (
 	"github.com/openshift-eng/openshift-tests-extension/pkg/cmd"
 	"github.com/openshift-eng/openshift-tests-extension/pkg/extension"
 	e "github.com/openshift-eng/openshift-tests-extension/pkg/extension"
-	"github.com/openshift-eng/openshift-tests-extension/pkg/extension/extensiontests"
+	et "github.com/openshift-eng/openshift-tests-extension/pkg/extension/extensiontests"
```
[the remark this answers is not in the document]
Comment (stbenjam) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:16]: Plans are to move origin to OTE itself for its own internal tests.  It'll use the same code we're using here and annotations will go away there too. 

We have no plans to require all tests to use extensions, adding tests to origin is still a valid path and will be for the foreseeable future.  But annotate will go away.

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go (hunk 2 above)
Comment (bertinatto) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go]: > What are these jobs used for? Is it even needed anymore with extensions? If you open a kube bump PR, any new tests are automatically run in the presubmits so why do we need this separate script?

Good question. I think those jobs have been around since the beginning, so I can't say for certain what the original intentions were, but I can guess a few reasons.

I believe it’s important that OCP passes upstream tests using the upstream binary and upstream images, as they were intended. While these tests may seem redundant, having some overlap is safer than missing a critical failure. Also, to obtain the kube certification, we need to run a different program that I believe uses a variation of the upstream binary and the upstream images.

I'm hesitant to remove these jobs. Any ideas how can keep them running with the incoming changes?

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go [hunk 6]
```diff
@@ -97,12 +97,54 @@ func main() {
...
+		et.NameContains("[Driver: aws]"),
+		et.NameContains("[Feature:StorageProvider]"),
+	}).Exclude(et.NoOptionalCapabilitiesExist())
+
+	specs.SelectAll([]et.SelectFunction{
+		// ovn-kubernetes does not support named ports
+		et.NameContains("NetworkPolicy"),
+		et.NameContains("named port"),
+	}).Exclude(et.NetworkEquals("OVNKubernetes"))
+
```
[the remark this answers is not in the document]
Comment (smg247) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:109]: I could, I was trying to start with the skips only just to be sure that the way I was doing it was correct. I guess that we have determined that with the back-and-forth here though, so no harm in doing them all in a single PR.

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go (hunk 2 above)
Comment (stbenjam) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go]: I think we could use a focus file https://onsi.github.io/ginkgo/#location-based-filtering to explicitly list all the tests we want to run, and use `k8s-test-ext` with environment flags or other filters to generate the list.

File: openshift-hack/cmd/k8s-tests-ext/environemnt-selectors.go
Comment (stbenjam) [inline openshift-hack/cmd/k8s-tests-ext/environemnt-selectors.go:1]: s/environemnt/environment/

File: openshift-hack/cmd/k8s-tests-ext/disabled-tests.go
Comment (stbenjam) [inline openshift-hack/cmd/k8s-tests-ext/disabled-tests.go:1]: Underscores are more 100x more common than dashes for go files in k/k.  Maybe the dashes are consistent in this directory because k8s-tests.go has a dash.

**Review by stbenjam — COMMENTED** [review]
I like it

File: openshift-hack/cmd/k8s-tests-ext/disabled_tests.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+			"[Driver: rbd]",             // OpenShift 4.x does not support Ceph RBD (use CSI instead)
+			"[Driver: ceph]",            // OpenShift 4.x does not support CephFS (use CSI instead)
+			"[Driver: gluster]",         // OpenShift 4.x does not support Gluster
+			"Volumes GlusterFS",         // OpenShift 4.x does not support Gluster
+			"GlusterDynamicProvisioner", // OpenShift 4.x does not support Gluster
+
+			// Skip vSphere-specific storage tests. The standard in-tree storage tests for vSphere
+			// (prefixed with "In-tree Volumes [Driver: vsphere]") are enough for testing this plugin.
+			// https://bugzilla.redhat.com/show_bug.cgi?id=2019115
+			"[sig-storage].*[Feature:vsphere]", //TODO: this rule is broken due to the regex, but it wouldn't pick anything up anyways, safe to remove?
```
Comment (smg247) [inline openshift-hack/cmd/k8s-tests-ext/disabled_tests.go:194]: Note the TODO here, what should we do about this one?

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bertinatto) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go]: Many of the existing rules are outdated and don't match any existing tests, so I think ` MustSelect`-like helper would be very helpful.

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go (hunk 2 above)
Comment (bertinatto) [reply openshift-hack/cmd/k8s-tests-ext/k8s-tests.go]: Can we use a skip file instead? `k8s-test-ext` would need to list the tests that need to be skipped, but based on https://github.com/openshift/kubernetes/pull/2215#discussion_r1989016427 I believe that's doable, correct?

File: openshift-hack/cmd/k8s-tests-ext/k8s-tests.go [hunk 8]
```diff
@@ -97,12 +97,19 @@ func main() {
...
 		}
 	})
 
+	specs = filterOutDisabledSpecs(specs)
+	addLabelsToSpecs(specs)
+
+	// EnvironmentSelectors added to the appropriate specs to facilitate including or excluding them
+	// based on attributes of the cluster they are running on
+	addEnvironmentSelectors(specs)
+
```
Comment (bertinatto) [inline openshift-hack/cmd/k8s-tests-ext/k8s-tests.go:112]: Looking much better, thanks!

File: openshift-hack/cmd/k8s-tests-ext/disabled_tests.go (hunk 7 above)
Comment (bertinatto) [reply openshift-hack/cmd/k8s-tests-ext/disabled_tests.go:194]: They existed in 4.15 [1], but seem to be gone in 4.16. Feel free to drop it. See? We need a `MustSelect` :D

[1] https://github.com/bertinatto/kubernetes/blob/ff493be70058d7324491442be6d57c92cfccb92d/test/e2e/storage/vsphere/persistent_volumes-vsphere.go#L37

Comment (smg247) [reply openshift-hack/cmd/k8s-tests-ext/disabled_tests.go:194]: I do think that will be useful, but can we do something like that in a followup?

Comment (bertinatto) [reply openshift-hack/cmd/k8s-tests-ext/disabled_tests.go]: Definitely

**Review by bertinatto — COMMENTED** [review]
/approve

**Review by bertinatto — COMMENTED** [review]
Things that I'd like to see in follow-ups:

1. Some mechanism to detect a rule that doesn't match any test
2. Some mechanism to continue running `k8s-*` jobs in this repository once the annotations are gone

**Review by bertinatto — COMMENTED** [review]
/remove-label backports/unvalidated-commits
/label backports/validated-commits
